src/openrct2/ride/TrackPaint.cpp

```diff
@@ -2184,6 +2185,164 @@ void track_paint(paint_session* session, uint8_t direction, int32_t height, cons
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 32, -8, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                            case 1:
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 32, 8, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 32, 4, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                            case 2:
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 32, -8, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 32, -4, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                        }
+                        break;
+                    case 3:
+                        switch (get_current_rotation())
+                        {
+                            case 2:
+                            case 3:
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 0, -16, height + zOffset, LIGHTFX_LIGHT_TYPE_LANTERN_3);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 8, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, -8, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                            case 0:
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 8, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, 4, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                            case 1:
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, -8, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                lightfx_add_3d_light_magic_from_drawing_tile(
+                                    session->MapPosition, -4, -32, height, LIGHTFX_LIGHT_TYPE_SPOT_1);
+                                break;
+                        }
+                        break;
+                }
```

Comment:
I believe it would improve legibility if you split this off to its own function.